storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerAppendInfo.java

```diff
@@ -87,24 +87,57 @@ public long producerId() {
     }
 
     private void maybeValidateDataBatch(short producerEpoch, int firstSeq, long offset) {
-        checkProducerEpoch(producerEpoch, offset);
+        // Default transaction version 0 is passed for data batches.
+        checkProducerEpoch(producerEpoch, offset, (short) 0);
         if (origin == AppendOrigin.CLIENT) {
             checkSequence(producerEpoch, firstSeq, offset);
         }
     }
 
-    private void checkProducerEpoch(short producerEpoch, long offset) {
-        if (producerEpoch < updatedEntry.producerEpoch()) {
-            String message = "Epoch of producer " + producerId + " at offset " + offset + " in " + topicPartition +
-                    " is " + producerEpoch + ", " + "which is smaller than the last seen epoch " + updatedEntry.producerEpoch();
+    /**
+     * Validates the producer epoch for transaction markers based on the transaction version.
+     * 
+     * <p>For Transaction Version 2 (TV2) and above (KIP-1228), the coordinator always increments
+     * the producer epoch by one before writing the final transaction marker. This establishes a
+     * clear invariant: a valid TV2 marker must have an epoch strictly greater than the producer's
+     * current epoch at the leader. Any marker with markerEpoch <= currentEpoch is a late or duplicate
+     * marker and must be rejected to prevent conflating multiple transactions under the same epoch,
+     * which would threaten exactly-once semantics (EOS) guarantees.
+     * 
+     * <p>For legacy transaction versions (TV0/TV1), markers were written with the same epoch as
+     * the transactional records, so we accept markers when markerEpoch >= currentEpoch. This
+     * preserves backward compatibility but cannot distinguish between active and stale markers.
+     * 
+     * @param producerEpoch the epoch from the transaction marker
+     * @param offset the offset where the marker will be written
+     * @param transactionVersion the transaction version (0/1 = legacy, 2 = TV2)
+     */
+    private void checkProducerEpoch(short producerEpoch, long offset, short transactionVersion) {
+        short current = updatedEntry.producerEpoch();
 
-            if (origin == AppendOrigin.REPLICATION) {
-                log.warn(message);
-            } else {
-                // Starting from 2.7, we replaced ProducerFenced error with InvalidProducerEpoch in the
-                // producer send response callback to differentiate from the former fatal exception,
-                // letting client abort the ongoing transaction and retry.
-                throw new InvalidProducerEpochException(message);
+        if (transactionVersion >= 2) {
+            if (producerEpoch <= current) {
+                String message = "Reject late/dup TV2 marker: markerEpoch=" + producerEpoch + " <= currentEpoch=" + current +
```

Comment:
Redundant code here, we could just do `boolean invalidEpoch = (transactionVersion >= 2) ? (producerEpoch <= current) : (producerEpoch < current);` and then `if (invalidEpoch)`.